language/tools/move-bytecode-utils/src/layout.rs

```diff
@@ -158,55 +174,115 @@ impl<'a, T: GetModule> SerdeLayoutBuilder<'a, T> {
         module_id: &ModuleId,
         name: &Identifier,
         type_arguments: &[Format],
-        shallow: bool,
     ) -> Result<Format, T::Error> {
         // build a human-readable name for the struct type. this should do the same thing as
         // StructTag::display(), but it's not easy to use that code here
-        let struct_key = if type_arguments.is_empty() {
-            format!("{}::{}", module_id, name)
+
+        let declaring_module = self
+            .module_resolver
+            .get_module_by_id(module_id)?
+            .expect("Failed to resolve module");
+        let def = declaring_module
+            .borrow()
+            .find_struct_def_by_name(name)
+            .unwrap_or_else(|| {
+                panic!(
+                    "Could not find struct named {} in module {}",
+                    name,
+                    declaring_module.borrow().name()
+                )
+            });
+        let normalized_struct = Struct::new(declaring_module.borrow(), def).1;
+        assert_eq!(
+            normalized_struct.type_parameters.len(),
+            type_arguments.len(),
+            "Wrong number of type arguments for struct"
+        );
+
+        let generics: Vec<String> = type_arguments
+            .iter()
+            .zip(normalized_struct.type_parameters.iter())
+            .filter_map(|(type_arg, type_param)| {
+                if self.config.ignore_phantom_types && type_param.is_phantom {
+                    // do not include phantom type arguments in the struct key, since they do not affect the struct layout
+                    None
+                } else {
+                    Some(print_format_type(type_arg))
+                }
+            })
+            .collect();
+        let struct_key_module = if self.config.omit_addresses {
+            format!("{}", module_id.name())
+        } else {
+            format!(
+                "{}{}{}",
+                module_id.address(),
+                self.config.separator.as_deref().unwrap_or("::"),
+                module_id.name()
+            )
+        };
+        let struct_key = if generics.is_empty() {
+            format!(
+                "{}{}{}",
+                struct_key_module,
+                self.config.separator.as_deref().unwrap_or("::"),
+                name
+            )
         } else {
-            let generics: Vec<String> = type_arguments.iter().map(print_format_type).collect();
-            format!("{}::{}<{}>", module_id, name, generics.join(","))
+            format!(
+                "{}{}{}{}{}{}",
```

Comment:
Maybe this whole pattern section would be easier to read with `write!`?

e.g. 
```
let mut struct_key = String::new();
if !self.config.omit_addresses {
  write!(f, "{}{}", module_id.address(), self.config.separator.as_deref().unwrap_or("::"));
}
write!(f, "{}{}{}", module_id.name(), self.config.separator.as_deref().unwrap_or("::"), name);
if !generics.is_empty() {
   ... 
}
```
This way you would get a lot more code sharing

Reply:
Good suggestion, that is much more concise

Reply:
I don't think string can be used with `write` unfortunately. I think `+=` might work though?

Reply:
It works with `string`.